apps/src/lib/node/ledger/shell.rs

```diff
@@ -266,48 +270,89 @@ impl Shell {
-                    tracing::info!(
-                        "all VPs accepted apply_tx storage modification {:#?}",
-                        result
-                    );
-                    self.write_log.commit_tx();
-                } else {
-                    tracing::info!(
-                        "some VPs rejected apply_tx storage modification {:#?}",
-                        result.vps_result.rejected_vps
-                    );
-                    self.write_log.drop_tx();
-                    response.code = 1;
+    /// Check the fees and signatures of the fee payer for a transaction
+    pub fn process_proposal(
+        &mut self,
+        _req: shim::request::ProcessProposal,
+    ) -> shim::response::ProcessProposal {
+        Default::default()
+    }
+
+    pub fn revert_proposal(
+        &mut self,
+        _req: shim::request::RevertProposal,
+    ) -> shim::response::RevertProposal {
+        Default::default()
+    }
+
+    pub fn extend_vote(
+        &mut self,
+        _req: shim::request::ExtendVote,
+    ) -> shim::response::ExtendVote {
+        Default::default()
+    }
+
+    /// Validate and apply transactions.
+    pub fn finalize_block(
+        &mut self,
+        req: shim::request::FinalizeBlock,
+    ) -> Result<shim::response::FinalizeBlock> {
+        let mut response = shim::response::FinalizeBlock::default();
+        for tx in &req.txs {
+            let mut tx_result =
+                Event::new_tx_event(EventType::Applied, tx, req.height);
+            match protocol::apply_tx(
+                tx,
+                &mut self.gas_meter,
+                &mut self.write_log,
+                &self.storage,
+            )
+            .map_err(Error::TxApply)
+            {
+                Ok(result) => {
+                    if result.is_accepted() {
+                        tracing::info!(
+                            "all VPs accepted apply_tx storage modification \
+                             {:#?}",
+                            result
+                        );
+                        self.write_log.commit_tx();
+                        tx_result["code"] = "0".into();
+                    } else {
+                        tracing::info!(
+                            "some VPs rejected apply_tx storage modification \
+                             {:#?}",
+                            result.vps_result.rejected_vps
+                        );
+                        self.write_log.drop_tx();
+                        tx_result["code"] = "1".into();
+                    }
+                    tx_result["gas_used"] = result.gas_used.to_string();
+                    tx_result["info"] = result.to_string();
+                }
+                Err(msg) => {
+                    tx_result["gas_used"] = self
+                        .gas_meter
+                        .get_current_transaction_gas()
+                        .to_string();
+                    tx_result["info"] = msg.to_string();
```

Comment:
This branch (occurs when there's a runtime error in a tx) is missing the response code. Let's maybe use `"2"` for it
```suggestion
                    tx_result["info"] = msg.to_string();
                    tx_result["code"] = "2".into();
```